nav2-usability/scenario-code/nested-routing/router.dart

```diff
@@ -0,0 +1,309 @@
+  Future<AppRoutePath> parseRouteInformation(
+      RouteInformation routeInformation) async {
+    var uri = Uri.parse(routeInformation.location!);
+    if (uri.pathSegments.isEmpty) {
+      return NewBooksRoutePath();
+    }
+    if (uri.pathSegments.length == 1) {
+      if (uri.pathSegments[0] == 'books') return NewBooksRoutePath();
+      if (uri.pathSegments[0] == 'settings') return SettingsRoutePath();
+    }
+
+    if (uri.pathSegments.length == 2) {
+      if (uri.pathSegments[1] == 'new') return NewBooksRoutePath();
+      if (uri.pathSegments[1] == 'all') return AllBooksRoutePath();
+    }
+
+    return NewBooksRoutePath();
+  }
+
+  @override
+  RouteInformation restoreRouteInformation(AppRoutePath path) {
+    late final location;
+
+    if (path is NewBooksRoutePath) {
+      location = '/books/new';
+    } else if (path is AllBooksRoutePath) {
+      location = '/books/all';
+    } else if (path is SettingsRoutePath) {
+      location = '/settings';
+    }
+
+    return RouteInformation(location: location);
+  }
+}
+
+class BookRouterDelegate extends RouterDelegate<AppRoutePath>
+    with ChangeNotifier, PopNavigatorRouterDelegateMixin<AppRoutePath> {
+  @override
+  final GlobalKey<NavigatorState> navigatorKey;
+  final AppState _appState = AppState();
+
+  BookRouterDelegate() : navigatorKey = GlobalKey<NavigatorState>() {
+    _appState.addListener(() => notifyListeners());
+  }
+
+  @override
+  Future<void> setNewRoutePath(AppRoutePath path) async {
+    if (path is NewBooksRoutePath) {
+      _appState.page = AppPage.newBooks;
+    } else if (path is AllBooksRoutePath) {
+      _appState.page = AppPage.allBooks;
+    } else if (path is SettingsRoutePath) {
+      _appState.page = AppPage.settings;
+    }
+  }
+
+  @override
+  void dispose() {
+    _appState.dispose();
+    super.dispose();
+  }
+
+  @override
+  AppRoutePath get currentConfiguration {
+    if (_appState.page == AppPage.newBooks) {
+      return NewBooksRoutePath();
+    } else if (_appState.page == AppPage.allBooks) {
+      return AllBooksRoutePath();
+    } else if (_appState.page == AppPage.settings) {
+      return SettingsRoutePath();
+    }
+    return AppRoutePath();
+  }
+
+  void _handleTabSelected(int idx) {
+    if (idx == 0) {
+      _appState.page = AppPage.newBooks;
+    } else {
+      _appState.page = AppPage.allBooks;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    late final Widget innerScreen;
+    if (_appState.page == AppPage.settings) {
+      innerScreen = SettingsScreen();
+    } else {
+      innerScreen = BooksScreen(
+        appPage: _appState.page,
+        onTabSelected: _handleTabSelected,
+      );
+    }
+
+    final bottomBarIndex =
+    _appState.page == AppPage.allBooks || _appState.page == AppPage.newBooks
+        ? 0
+        : 1;
+    return Navigator(
+      key: navigatorKey,
+      pages: [
+        MaterialPage(
+          key: ValueKey('BooksScreen'),
+          child: AppScaffold(
```

Comment:
We can move AppScaffold outside of the navigator? since the appbar and bottomnavigation bar are static content

Reply:
I think the BottomNavigationBar threw a "No Overlay widget found" when I tried this.

Reply:
You need to wrap `AppScaffold` in another `Navigator`, as @nguyenxndaidev mentionned [here](https://github.com/flutter/uxr/pull/34#issuecomment-813396855)

Reply:
You don't need to have another navigator, just wrap a overlay
```dart
Overlay(
        initialEntries: [ OverlayEntry(builder: (_) =>AppScaffold(
            currentIndex: bottomBarIndex,
            onIndexChanged: (idx) {
              if (idx == 0) {
                _appState.page = AppPage.newBooks;
              } else {
                _appState.page = AppPage.settings;
              }
            },
            child: Navigator(
              key: navigatorKey,
              pages: [
                MaterialPage(
                  key: ValueKey('BooksScreen'),
                  child: innerScreen,
                ),
              ],
              onPopPage: (route, result) {
                return route.didPop(result);
              },
            )
        ))]
    );
```

Reply:
The reason I think we should move the AppScaffold out is that we want to emphasis on the ability to build static content outside of the `router-outlet`

Reply:
one thing of using the overlay entry is that you will have to wire the setState into the OverlayEntry because the
Overlay.initialEntries will not update when it changes.

Reply:
For example something like this https://gist.github.com/chunhtai/a21f09b647ecc782e9ee0ca9f95f2282

All the Overlay annoyances is because of the static content needs to be below the overlay. If we use the another widget to replace BottomNavigationBar, we don't have to deal with this. However, I think this is still a good example as to see how other library would support this case. If user want to build static content that requires Overlay/Navigator, does the library put the responsibility to the user? or do they just wrap a top level navigator by default.

Reply:
Moving the AppScaffold out sounds good. Can you clarify what you mean by "wire the setState into the OverlayEntry"?

Reply:
you can see the gist i posted in the previous comment. So the problem with the Overlay widget is that it is imperative. so rebuilds the overlay with different initialOverlayEntires will not rebuilds the overlay entry, you need to cache the overlay entry yourself and call markNeedsBuild manually when you want it to rebuild.

Reply:
That seems to work for most cases, but now I'm seeing an issue when the user manually edits the route from `/books/new` to `/books/all`.

Reply:
I will take a look later this afternoon

Reply:
No, I meant to keep the TabBarController animation in this case, and use a fade animation between `books/*` and `settings/`